client/Categorifier/Client/Internal.hs

```diff
@@ -16,17 +17,53 @@ import qualified Categorifier.Common.IO.Exception as Exception
 import Categorifier.Duoidal (Parallel (..), traverseD)
 import Categorifier.Duoidal.Either (noteAccum)
 import qualified Categorifier.TH as TH
+import Control.Monad ((<=<))
 import Data.Bifunctor (Bifunctor (..))
 import Data.Constraint (Dict (..))
 import Data.Foldable (foldl', toList)
 import Data.List.NonEmpty (NonEmpty (..))
-import qualified Data.List.NonEmpty as NE
-import Data.Maybe (listToMaybe)
+import Data.Maybe (isJust, listToMaybe)
 import Data.Semigroup (Any (..))
 import Data.Tuple.Extra (fst3, snd3, thd3)
 import Data.Void (Void)
 import PyF (fmt)
 
+-- | The "standard representation" that forms an isomorphism with the underlying type. This type
+--   should only modify the current type (not the types contained within this one), and should
+--   consist only of `Either`, `Void`, `()`, `(,)` (strictly 2-tuples), and `Dict`.
```

Comment:
As far as I can remember, this is not a requirement, right? For `data Foo = Foo Bar Baz`, you can define `Rep Foo` as either `(Bar, Baz)` or `(Rep Bar, Rep Baz)`, although there might be some performance difference (and I don't remember which one is faster).

Also this sentence seems contradictory - "should only modify the current type" indicates `Rep Foo = (Bar, Baz)`, but "consist only of ..." suggests `Rep Foo = (Rep Bar, Rep Baz)`.

Reply:
You're right. I'll rephrase this to be clearer. The desired type is `Rep Foo ~ (Bar, Baz)`. In the pieces that get replaced, you should only use `Either`, `Void`, `()`, `(,)`, and `Dict`, because those are the only things we special-case in `categorify`. I think recursing deeper (like `(Rep Bar, Rep Baz)`) might actually cause problems, but I can't think of what off the top of my head.

Reply:
Let me know what you think of the new phrasing.

Reply:
Looks much better.

> making the structure as shallow as possible helps with performance

Is this true? I don't remember but I thought it's the opposite. If you use 3-tuples in `Rep`, then it would take another iteration to get to 2-tuples, so I imagine it's faster to use 2-tuples directly.

Reply:
So, I mention above this part that you should only use 2-tuples. As shallow as possible means that

`((a, b), (c, d))`
is better than
`(a, (b, (c, d)))`

I'll rephrase some more.

Reply:
Ah, right. Looks good.